### ISSUE TITLE
Support for Windows Phone 8.1

### ISSUE THREAD

**Issue by Aurora12** [adamcaudill/libsodium-net#96]
Hello,

I want to use libsodium-net in a Windows Store Universal App. This type of projects was introduced this spring and features sharing code between the mobile and desktop parts of the project. This works starting with Windows 8.1 and uses a restricted subset of .NET apis (somewhat even more restricted than Windows Phone 8 and earlier versions had). NuGet package for this kind of projects should target 'WindowsPhoneApp,Version=v8.1'.

There are two problems here:
  1. .NET apis: 
    * **AppDomain** class is unavailable and there is no loading dlls at runtime as far as I know. So, you can not use **Sodium.DynamicInvoke**.
    * **System.Security.Cryptograph**y is now **Windows.Security.Cryptography**
    * **Array.Copy()** last parameter is no longer long, it's int.
    * **ProtectedMemory** and **MemoryProtectionScope** are to be somehow replaced with Windows.Security.Cryptography.DataProtection.DataProtectionProvider. Here's the full list of replacements: http://msdn.microsoft.com/en-us/library/windows/apps/hh464945.aspx
  2. I need to **statically** link sodium library to Windows Phone part of the project. Current libsodium-1.0.1 compiles for a variety of platforms, but not for Windows Phone ARM (let alone 8.1, for that matter). On MSDN forums they suggest that you should create a C++ Static Library or Windows Runtime component and compile the code with Visual Studio (so you can use the result in a Universal Store App). 

While I could possibly cope with the first part, I have no clues as to what to do with the second one. Any ideas?

**Comment by jedisct1** [comment]
That can be worked around. What is the proper way to detect Windows 8.1 Mobile?

**Comment by jedisct1** [comment]
Maybe something like https://github.com/jedisct1/libsodium/commit/820c84ab894667f2ad0cef4e9cf849bee0a1fe71 can do the trick.

Is `timeval` actually not being defined anywhere on this target? It's safe to remove the `randombytes_salsa20` folder from the project if you don't use it, but `timeval` being unavailable is quite surprising.

**Comment by Impyy** [comment]
I'm sorry for jumping in on the discussion like this but I figured this wasn't worth creating a separate issue.
I'm trying to compile libsodium as a runtime component too but I'm getting some errors. jedisct1/libsodium@820c84a fixed some of the errors Aurora12 was having but I'm still stuck with these:

```
error C2719: '_A': formal parameter with __declspec(align('16')) won't be aligned (src\libsodium\crypto_pwhash\scryptsalsa208sha256\sse\pwhash_scryptsalsa208sha256_sse.c)	C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\xmmintrin.h	186	1	libsodium
error C2719: '_B': formal parameter with __declspec(align('16')) won't be aligned (src\libsodium\crypto_pwhash\scryptsalsa208sha256\sse\pwhash_scryptsalsa208sha256_sse.c)	C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\xmmintrin.h	186	1	libsodium
error C2719: '_A': formal parameter with __declspec(align('16')) won't be aligned (src\libsodium\crypto_pwhash\scryptsalsa208sha256\sse\pwhash_scryptsalsa208sha256_sse.c)	C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\xmmintrin.h	187	1	libsodium
error C1083: Cannot open include file: 'wincrypt.h': No such file or directory
```
(about 100+ of C2719)

**Comment by jedisct1** [comment]
To work around the first issue, remove `|| defined(_MSC_VER)` in `pwhash_scryptsalsa208sha256_sse.c` and `crypto_scrypt-common.c`.

The absence of `wincrypt.h` is more concerning. Try to remove the `#include <wincrypt.h>` statement in `utils.c` and see how it goes.

**Comment by Impyy** [comment]
Forget about wincrypt.h not being found. It exists, I just wasn't including the right directory, sorry about that.
After that I got an error about the fact that advapi32.lib was missing, I fixed that by adding ``C:\Program Files (x86)\Windows Kits\8.1\Lib\winv6.3\um\arm`` to the list of library directories

Now this error pops up:
```
error LNK2019: unresolved external symbol __cpuid referenced in function _cpuid runtime.obj
```

Thanks for your help so far.

**Comment by jedisct1** [comment]
Replace `#ifdef _MSC_VER` with `#if 0` in `runtime.c`.

**Comment by Impyy** [comment]
That's it, it compiled. Thank you for your help!

**Comment by jedisct1** [comment]
Woo-ooh!

**Comment by Aurora12** [comment]
@jedisct1 Thanks for looking into the issue!

I excluded randombytes_salsa20_random.c from the project since I don't use it, and this eliminated `timeval` errors. Windows Store apps do not support Winsock APIs (unlike Windows Phone 8 apps), which is the reason for timeval not being there, as far as I know.

The project doesn't compile still, looks like the fix is not effective.
This is what VS adds to definitions in  WP8.1 project:
`PSAPI_VERSION=2;WINAPI_FAMILY=WINAPI_FAMILY_PHONE_APP;_UITHREADCTXT_SUPPORT=0;%(PreprocessorDefinitions)`
I guess this is the way to detect WP platform.
Could you please modify your fix?

**Comment by Aurora12** [comment]
@jedisct1 I added a couple of lines in utils.c and runtime.c just above your previous additions.
Did so according to http://blogs.msdn.com/b/chuckw/archive/2012/09/18/dual-use-coding-techniques-for-games-part-3.aspx

```c
#ifdef _WIN32
# define WINAPI_DESKTOP
# if defined(WINAPI_FAMILY) && (WINAPI_FAMILY == WINAPI_FAMILY_PHONE_APP)
#  undef WINAPI_DESKTOP
# elif defined(WINAPI_FAMILY_ONE_PARTITION) && defined(WINAPI_FAMILY_DESKTOP) && !WINAPI_FAMILY_ONE_PARTITION(WINAPI_FAMILY_DESKTOP)
#  undef WINAPI_DESKTOP
# endif
#endif
```
Now the project compiles! 
Currently I'm testing it to see if all works fine on devices.

**Comment by Aurora12** [comment]
Oh, well, shit got real. While the project does compile for windows phone 8.1, it still does not compile for Windows Store project, which is what I need in the end. I made some corrections to code, but got stuck with several things.

Corrections I've made:
* __inline instead of inline
* disabled /ZW and /Yu for all C files 
* #include <windows.security.cryptography.h> instead of # include <wincrypt.h>

Errors I have now:
* C2099: Initializer is not a constant in randombytes.c on line:
```
static const randombytes_implementation *implementation =
    &randombytes_sysrandom_implementation;
```
* C4146: unary minus operator applied to unsigned type, result still unsigned in fe_cswap_curve25519_ref10.c, fe_cmov.c, randombytes_sysrandom.c in expressions like b = -b; with b being unisigned int.

I'd appreciate any hints as to what to do with this stuff (why does it compile for other platforms anyway?)

**Comment by Impyy** [comment]
@Aurora12 The former can be solved by defining SODIUM_STATIC. The latter can be solved by disabling SDL checks.

**Comment by jedisct1** [comment]
Can you check if the current libsodium code in git properly works for you in a universal app (or at least on Windows Phone)?

**Comment by Aurora12** [comment]
@jedisct1 Will give it another try this week! For several months I've been using js implementation of nacl through web view, which is of course ridiculous (not to say hilarious), but I had more important business logic problems to solve. Now I need to get it really working both in direct and in stream mode. Will update this thread on the results.

**Comment by Aurora12** [comment]
@jedisct1 So, I gave it another try and was partly successfull.

I was able to compile libsodium for ARM and x86/x64 in a Windows Runtime Component project for Universal Apps (targeting Windows 8.1 and Windows Phone 8.1). I had to do quite a lot of stuff to get it working (I can list it if you like), and had to write a wrapper class that exposes methods to C# using C++/CX types. I then created a test project of a Universal Store App in C# and tried to use the library. 

Works fine in Windows 8.1 part of the project (I tried boxing and signing). 

Doesn't work for Windows Phone 8.1 part. It looks like sodium fails to even initialize: the first reference to C++ wrapper class yields exception "Module not loaded" with no details as to which module failed. I tried launching x86 build in phone emulator and ARM build on device and they yield the same result.

What I do now is I'm adding code files to libsodium sources one by one to see when it stops loading in winphone. I started with just core.c, utils.c and runtime.c. It's rather hard since I'm not all that comfortable with C or C++ projects.

**Comment by Aurora12** [comment]
@jedisct1 @adamcaudill Looks like windows phone 8.1 stops to load library assemblies once I include randombytes_sysrandom.c into project. The library compiles and I commented out a couple of lines in core.c and utils.c, so the code is actually not triggered. Once randombytes_sysrandom.c is there, library fails to load on windows phone 8.1. Any ideas?

1>  randombytes.c
1>  randombytes_sysrandom.c
1>  core.c
1>  runtime.c
1>  utils.c
1>  version.c

These are the only files from libsodium I compile now.

**Comment by Aurora12** [comment]
@jedisct1 @adamcaudill Looks like `RtlGenRandom` is not available on wp 8.1, so this doesn't work:
```
#ifdef _WIN32
# include <windows.h>
# define RtlGenRandom SystemFunction036
# if defined(__cplusplus)
extern "C"
# endif
BOOLEAN NTAPI RtlGenRandom(PVOID RandomBuffer, ULONG RandomBufferLength);
# pragma comment(lib, "advapi32.lib")
#endif
```

As I mentioned in previous comment, libsodium fails to load if I compile randombytes_sysrandom.c. But when I comment out the following lines, the library loads (but is not working, obviously):
```
if (! RtlGenRandom((PVOID) buf, (ULONG) size)) {
    abort(); /* LCOV_EXCL_LINE */
}
```

https://msdn.microsoft.com/en-us/library/windows/desktop/aa387694%28v=vs.85%29.aspx
This page suggessts that we should use `CryptGenRandom` instead.

**Comment by jedisct1** [comment]
Thanks for pinpointing the culprit. Is it the only remaining thing that doesn't work on WP 8.1?

`CryptGenRandom()` not being available is a bummer. 

`Windows::Security::Cryptography::CryptographicBuffer::GenerateRandom()` looks like a way to do it, but is it available as a C function as well? Mind grepping for "generaterandom` in the available header files?

**Comment by Aurora12** [comment]
@jedisct1 There is `wincrypt.h` file in Windows Kit 8.1, and it does contain `CryptGenRandom()`. But almost all of the code there, including CryptGenRandom(), is inside `#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)`, so it's desktop only. And Windows Phone Kit 8.1 does not contain wincrypt.h at all.

The only other file to contain generaterandom something is `schannel.h` which is in Windows Kit and contains `Public Definitions for SCHANNEL Security Provider`. 
```// Pre NT4.0 SP2 calls.  Call CAPI1 or CAPI2
// to get the same functionality instead.
VOID
WINAPI
SslGenerateRandomBits(
    PUCHAR      pRandomData,
    LONG        cRandomData
    );```

**Comment by Aurora12** [comment]
I wonder if I could use another approach. What if I generate random bytes in my wrapper class and pass them to libsodium methods thus avoiding using `randombytes_buf()` altogether? This would be awkward, yes, and would require writing wrapper methods inside libsodium, but I need it working at long last. :)

**Comment by Aurora12** [comment]
@jedisct1 Just discovered one another problem. 
```
#define SecureZeroMemory RtlSecureZeroMemory
```
resides in WinBase.h under
```
#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP)
```
So it's not accessible in `sodium_memzero()` on WP.

**Comment by Aurora12** [comment]
@jedisct1 There's also some good news.
I wrote the following and compiled it with /ZW.
```
// h:
extern "C" int GetRandomBytes(unsigned char *buffer, int length);

// cpp:
#include "pch.h"
#include "Extern.h"

using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation::Collections;
using namespace Windows::Security::Cryptography;
using namespace Windows::Storage::Streams;

extern "C" int GenerateRandomBytes(unsigned char *bytes, unsigned int length)
{
	IBuffer^ buffer = CryptographicBuffer::GenerateRandom(length);
	DataReader^ reader = DataReader::FromBuffer(buffer);
	reader->ReadBytes(ArrayReference<unsigned char>(bytes, buffer->Length));
	return 1;
}
```

I then replaced lines with `RtlGenRandom` in `randombytes_sysrandom.c` with the following:
```
if (!GenerateRandomBytes((unsigned char*)buf, (unsigned int)size)) {
    abort();
}
```

It seems to work fine on Windows 8.1 part of the project. The library still fails to load, but, apparently, the problem with random generator is fixed.

**Comment by jedisct1** [comment]
As a cleaner alternative to patching `randombytes_sysrandom.c`, you can register a custom implementation: http://doc.libsodium.org/advanced/custom_rng.html

Or, maybe even cleaner, declare a `SystemFunction036()` polyfill.

Still, that's fairly annoying. There has to be a way to do that in C.

**Comment by Aurora12** [comment]
@jedisct1 I totally agree. I left it for now as I'm still investigating the reasons the library fails to load. The problem with `SecureZeroMemory` I mentioned above may be a false alarm. It's quite difficult to debug this when running a winrt project.

**Comment by jedisct1** [comment]
`SecureZeroMemory()` not being available is not a big deal; the portable implementation is totally fine with all versions of Visual Studio so far.

But not having a PRNG is a showstopper.

**Comment by Aurora12** [comment]
@jedisct1 I finally got libsodium working in my universal store project. Here's the list of modifications to libsodium sources I had to make:
# Replace `static include` with `static __include` in all files
# Replace calls to `RtlGenRandom` with calls to a function (a couple of comments above) that  internally uses winrt apis and is compiled with /ZW. Alternativel, define `SystemFunction036` as you sugested above.  This affects `randombytes_sysrandom.c` and `randombytes_salsa20_random.c`. 
# Undefine WINAPI_DESKTOP symbol for windows phone (a patch to runtime.c from the beginning of the thread).

**Comment by GregorSpagnolo** [comment]
@Aurora12 how did you linked C -> C++ code?
In C project I created a cpp file and tried to compile as ZW. When hitting compile I receive the following error:
LNK2038	mismatch detected for 'vccorlib_lib_should_be_specified_before_msvcrt_lib_to_linker': value '1' doesn't match value '0' in MSVCRTD.lib(app_appinit.obj)

**Comment by Aurora12** [comment]
@GregorSpagnolo Here's what I did:

1. Created a project from template Visual C++ : Store Apps : Universal Apps : Windows Runtime Component (Universal Apps) (VS2013). 
2. Added libsodium/src to `Shared` project
3. Removed `Consume Winrt Extensions (/ZW)` and `Use Precompiled Headers (/Yu)` from all *.c files.
4. Copied properties from VS solution included in libsodium distribution (preprocessor symbols, disabled warnings)
5. Added reference to `Microsoft Visual C++ 2013 Runtime Package for Windows Phone` to Windows Phone project.
6. Made fixes to C code listed in previous comment.

I believe, you should be able to compile and initialize library with this setup.

**Comment by GregorSpagnolo** [comment]
@jedisct1 Ok finally it works. You can compile entire LibSodium source code for Win10 mobile.

**Comment by Aurora12** [comment]
@GregorSpagnolo what steps did you have to follow to make it work? The same I listed above? Did you have to change the source code of libsodium as I did?

**Comment by GregorSpagnolo** [comment]
@Aurora12 Yes I followed your steps.

**Comment by luzianscherrer** [comment]
@Aurora12 / @jedisct1 Did you finally get a reliable, working libsodium-net for your Windows Phone project? I'm also in need of this and I was wondering whether you would share the library that you've built.

**Comment by globeport** [comment]
@Aurora12 / @jedisct1 I'd also really appreciate a working library for Win 10 mobile/uwp. Do you have a working build you can publish? Thanks

**Comment by globeport** [comment]
I can confirm that it is possible to build libsodium for UWP following the advice given above including code edits.
NB 
The following flags needs to be added to the linker command line:

`/nodefaultlib:vccorlibd /nodefaultlib:msvcrtd vccorlibd.lib msvcrtd.lib `

And also make sure the pre-processor definitions include SODIUM_DLL_EXPORT

Some alterations are required to libsodium-net. 

1. Delete references to System.Security.Cryptography
2. Edit class GenericHashAlgorithm so that it no longer inherits from HashAlgorithm
3. Rework KeyPair to use DataProtectectionProvider in place of ProtectedMemory for Protect and Unprotect functions. These operations need to be async.
4. Remove the DynamicInvoke/LazyInvoke classes and all code references to them. Instead apply DllImport directives to each delegate in the SodiumLibrary class. For each delegate replace its declaration 'internal delegate' with 'internal static extern'

I've probably missed something here but I think those are the main steps to get libsodium-net up and running on UWP. Havent tested everything yet but so far so good.

**Comment by globeport** [comment]
However, alas, despite the above I am getting a compiler error integrating libsodium-net into my store app when in release mode with .net native tool chain enabled (compiles fine without). Does anyone have any ideas?

Internal compiler error: Object reference not set to an instance of an object.

**Comment by Aurora12** [comment]
@globeport Actually, I'm not using libsodium-net. I took libsodium library code in C, put it in a WinRT component, made adjustments as pointed out above, and wrote a C++/CX wrapper for the functions I need.

**Comment by globeport** [comment]
Ah ok. 

So, I have ended up replacing ServiceLibrary.cs (libsodium.net) with an equivalent C++/CX class.
I can now get everything to compile using the .net native tool chain. 

Also I had to change the pre-processor definitions to include SODIUM_STATIC instead of SODIUM_DLL_EXPORT.

**Comment by Aurora12** [comment]
Hello again! It's been quite a while since I last updated libsodium. Now I need Argon2, so the time has come to update the library. I worked with 1.0.3 previously, now I'm trying to use 1.0.11. Compiling in vs2013 as a Windows Runtime Component (Universal Apps).

I made some changes to source files, as discussed above, because I still don't know of a better way to make it compile. For instance, I still had to replace all "static inline" with "static __inline" which I reckon should be possible to avoid. And I can't figure out how to add a build of this type to libsodium/builds/msvc/, so for now I'm working in a fork: https://github.com/Aurora12/libsodium

Added a repo with test project and a c++/cx wrapper (far from being complete):
https://github.com/Aurora12/libsodium-universal

Currently I have these warnings (the first one is alarming):
> core.c(98): warning C4013: 'InitializeCriticalSection' undefined; assuming extern returning int
> utils.c(xxx): warning C4197: 'volatile const unsigned char *volatile ' : top-level volatile in cast is ignored

And a linker error (which wasn't there in the previous version I had):
> LINK : fatal error LNK1104: cannot open file 'advapi32.lib'

@jedisct1 Could you please take a look?

p.s. Actually, this topic is not related to libsodium-net, should I start it in libsodium?

**Comment by kangjoni76** [comment]
+1 on this

**Comment by charlesportwoodii** [comment]
@Aurora12 Thanks for your work figuring out the bulk of this.

For anyone else who might stumble upon/or need this. I've created a Universal Runtime Component that works for both desktop and mobile (windows 10): https://github.com/charlesportwoodii/libsodium-uwp.

It implements a sizable portion, but not all of the `libsodium` API (1.0.11-master), and thanks to @jedisct1 does not require any patching of the `libsodium` library. If there's a feature not yet implemented that you need let me know/or send a PR over.

**Comment by adamcaudill** [comment]
That's awesome news! Patching libsodium is the biggest issue I saw with this, glad to see it happen without patching.

**Comment by jedisct1** [comment]
Awesome job, Charles!

**Comment by globeport** [comment]
Fantastic, thanks @charlesportwoodii

**Comment by pdefreitas** [comment]
@charlesportwoodii did a fantastic job. His solution is good when you are working at C++ level, but if you are at a pure C level you need to patch libsodium.

**Comment by Aurora12** [comment]
Hello from the end of 2017! Windows Mobile platform is obsolete and we have charlesportwoodii's solution. I believe this issue should be closed.

**Comment by jedisct1** [comment]
R.I.P. Windows Mobile.

**Comment by noloader** [comment]
@jedisct1,

> R.I.P. Windows Mobile.

Lol... That's actually Windows Phone, not Windows Mobile. Windows Mobile is based on Windows CE and the predecessor to Windows Phone. Windows Mobile was not tangled up in UWP.

I still see a fair amount of Windows Mobile code in medical devices. Medical is a lot like FIPS. Once the code is approved is does not change. People still use it with the antique SDKs because it is too expensive to have medical devices re-certified with updated code.
